src/poolboy.erl

```diff
@@ -248,42 +244,35 @@ handle_info({'EXIT', Pid, _Reason}, State) ->
     #state{supervisor = Sup,
            monitors = Monitors,
            workers = Workers} = State,
-    ok = cancel_worker_reap(State, Pid),
-    case ets:lookup(Monitors, Pid) of
+    case ets:take(Monitors, Pid) of
         [{Pid, _, MRef}] ->
             true = erlang:demonitor(MRef),
-            true = ets:delete(Monitors, Pid),
             NewState = handle_worker_exit(Pid, State),
             {noreply, NewState};
         [] ->
-            case lists:member(Pid, Workers) of
+            case lists:keymember(Pid, 1, Workers) of
                 true ->
-                    W = lists:filter(fun (P) -> P =/= Pid end, Workers),
-                    {noreply, State#state{workers = [new_worker(Sup) | W]}};
+                    W = lists:keydelete(Pid, 1, Workers),
+                    {noreply, State#state{workers = [{new_worker(Sup), os:timestamp()} | W]}};
                 false ->
                     {noreply, State}
             end
     end;
-handle_info({reap_worker, Pid}, State)->
-    #state{monitors = Monitors,
-           workers_to_reap = WorkersToReap} = State,
-    true = ets:delete(WorkersToReap, Pid),
-    case ets:lookup(Monitors, Pid) of
-        [{Pid, _, MRef}] ->
-            true = erlang:demonitor(MRef),
-            true = ets:delete(Monitors, Pid),
-            NewState = purge_worker(Pid, State),
-            {noreply, NewState};
-        [] ->
-            NewState = purge_worker(Pid, State),
-            {noreply, NewState}
-    end;
+handle_info({rip_workers, TimerTTL}, State) ->
+    #state{overflow = Overflow,
+        overflow_ttl = OverTTL,
+        workers = Workers,
+        supervisor = Sup} = State,
+    Now = os:timestamp(),
```

Comment:
To avoid time warp, it's best to use erlang monotonic so that if the system clock changes underneath us we don't necessarily reap workers. See http://erlang.org/doc/apps/erts/time_correction.html